src/SourceLocator/StaticClassMapSourceLocator.php

```diff
@@ -29,7 +32,7 @@ public function __construct(
         $realPaths = array_map('realpath', $classMap);
 
         Assert::that($classMap)->all()->file();
-        Assert::that(array_keys($classMap))->all()->string()->notEmpty();
```

Comment:
`notEmpty` works on non-`string` values so we should keep the `string()` assertion

Reply:
All values here are guaranteed to be strings by type checks: it is not possible to trigger this assertion via a test anymore either

Reply:
To be clear, we had a test scenario for this, but refining the type declaration led me to remove the test too (and subsequently the assertion, as per mutation tests)

Reply:
This *IS* the type check: `@param` is not sufficient, nor is the `array` type declaration to be able to trust that the keys are all `string`s

Reply:
The type check happens statically, not at runtime.

As of current type declarations, it is impossible to write a test that type-checks (before runtime) and then triggers this assertion (mutation test requirement)

Reply:
It's not impossible:

```php
new StaticClassMapSourceLocator([1 => 'foo'], $astLocator)
```

Reply:
That's not respecting the declared type: it is wrong even without running it, which is where my argument lies: if I can prove it to be wrong, I don't need to execute it at all.

Reply:
Yes, understood; it IS wrong, and tools say it's wrong, but a *consumer* of this API can still do the wrong thing. You are trusting that the docs are being read or the consumer is using static analysis, which is not always true. Right now we have excellent static analysis tools (psalm/phpstan) but they only analyse, never enforce. My issue is that by removing the assertions, people can do the wrong thing. IMO it goes against the core argument of your Extremely Defensive PHP talk - the whole poka yoke thing (avoid mistakes / fool proofing) - you can only plug it in one way. If you remove the assertions, there is more than one way to plug this thing in, and one/some/all of them are wrong. Just because it SAYS it's wrong, doesn't mean you can't actually do things wrong.

Until we get to a point where psalm/phpstan can actively enforce type declarations like `array<string, string>` or `callable () : Foo`, AND that we can be sure that all consumers use these tools to respect a defined API, I really strongly thing that runtime assertions should be kept.

Reply:
> Yes, understood; it IS wrong, and tools say it's wrong, but a *consumer* of this API can still do the wrong thing.

Repeating something that we discussed in private, but I think this is a good foundation for dismissing issues as they come in. Much like "missing test case", "doesn't type-check" is also valid IMO.

 > IMO it goes against the core argument of your Extremely Defensive PHP talk - the whole poka yoke thing (avoid mistakes / fool proofing) - you can only plug it in one way.

Few things here:

 1. ["Extremely Defensive PHP"](https://github.com/Ocramius/extremely-defensive-php) is a talk from 2015, which I gave over 20 times at different conferences/events, and I still believe most of in, but it is very much in need for updates
 2. at the time that I wrote "Extremely Defensive PHP", tools like @phpstan and @vimeo /psalm and @doctrine /coding-standard were not existing, or in their infancy. @vimeo /psalm was first released in November 2016, @phpstan in July 2017, and I pushed @doctrine /coding-standard in January 2016
 3. the tools listed above supersede most of my (loudly touted) beliefs exposed in "Extremely Defensive PHP", and using them is very much equivalent to embracing [Poka-Yoke](https://en.wikipedia.org/wiki/Poka-yoke). It is always possible to skip them in some locations if you **really** know what you are doing, but things do indeed "click together" on "one true way of doing things"

Given the discussion so far, I think it is time for me to either design a new talk, or amend some stuff about it in a blogpost.

 > My issue is that by removing the assertions, people can do the wrong thing.

I agree that people are free to misuse the tool, but I am as free telling "this is not correct", and rejecting the issue (As mentioned above: "doesn't type-check" being equivalent to "cannot reproduce").

 > Until we get to a point where psalm/phpstan can actively enforce type declarations like `array<string, string>` or `callable () : Foo`, AND that we can be sure that all consumers use these tools to respect a defined API

This triggered a mental process on my end: as discussed in private, I could write a @vimeo /psalm plugin that acts on @composer `"post-install"` by running type-checks on any API surface that touches certain (configured) namespaces. I don't think that should affect this patch directly, but it is something that I will work on. It will be annoying as heck, but it will also be a massive improvement for the ecosystem, because opting out of type-checks will no longer be a viable choice.

 > I really strongly thing that runtime assertions should be kept.

I will see what I can do, but the current alternatives I see are:

 1. removing the added type declarations
 2. creating tests outside the `test/unit` (something like `test/api-misuse-errors`)

There is also the problem that assertions may fail to type-check for @phpstan or @vimeo /psalm as these tools catch up and start considering them (correctly) "redundant", which we will surely hit at some point, as type inference improves with the ecosystem tooling.

Reply:
@asgrim now that I released `roave/you-are-using-it-wrong`, can we re-discuss this? Do we need to add the dependency, or are we fine with the patch as-is?

As mentioned above, I can either remove the type declarations overall, or add tests that are outside static analysis checking.

Reply:
Ship it I reckon - maybe make a separate patch to add `you-are-using-it-wrong`? :shipit:

Reply:
Fair enuff